boot/modeenv.go

```diff
@@ -195,3 +196,47 @@ func splitModeenvStringList(v string) []string {
 func asModeenvStringList(v []string) string {
 	return strings.Join(v, ",")
 }
+
+// deepEquals compares two modeenvs to ensure they are textually the same. It
+// does not consider whether the modeenvs were read from disk or created purely
+// in memory. It also does not sort or otherwise mutate any sub-objects,
+// performing simple strict verification of sub-objects.
+func (m *Modeenv) deepEquals(m2 *Modeenv) bool {
+	// compare all the fields to be deep equals
+
+	// simple strings
+	if m.Mode != m2.Mode {
```

Comment:
Given that we're already importing reflect as indirect dependency, could we use reflect.DeepEqual() here? The only issue I think is that DeepEqual looks at unexported fields too, so it will get confused by `read` and `originRootDir`.

Reply:
yes I didn't want to use DeepEqual for precisely that reason, see https://play.golang.org/p/OQYlfWxvyGF. I think Michael's suggestion to use json.Marshal, etc. actually isn't too bad. I don't think we should be super concerned with performance right now so exporting to JSON shouldn't be too bad.